Base ScanFilterAndProjectOperator on WorkProcessors

sopel39 opened the pull request:
Part of: https://github.com/prestosql/presto/issues/49

This paves way for:
1. Late source pages materialization.
2. Reusing pages source (e.g: ORC) resources between pages.

sopel39 commented:
Benchmarks results:
```
BEFORE:
Benchmark                                                      (columnCount)  (dictionaryBlocks)  (positionsPerPage)   (type)  Mode  Cnt     Score    Error  Units
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              2               false                  32  varchar  avgt   50   291,888 ±  5,830  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              2               false                  32   bigint  avgt   50   129,592 ±  2,121  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              2               false                1024  varchar  avgt   50   145,449 ±  2,800  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              2               false                1024   bigint  avgt   50    15,233 ±  0,439  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              2                true                  32  varchar  avgt   50   202,027 ±  4,293  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              2                true                  32   bigint  avgt   50   136,185 ±  2,375  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              2                true                1024  varchar  avgt   50    51,044 ±  1,143  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              2                true                1024   bigint  avgt   50    13,019 ±  0,292  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              4               false                  32  varchar  avgt   50   473,653 ±  8,074  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              4               false                  32   bigint  avgt   50   209,716 ±  3,066  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              4               false                1024  varchar  avgt   50   222,964 ±  4,595  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              4               false                1024   bigint  avgt   50    23,868 ±  0,537  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              4                true                  32  varchar  avgt   50   340,191 ±  5,879  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              4                true                  32   bigint  avgt   50   224,762 ±  4,958  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              4                true                1024  varchar  avgt   50    86,607 ±  1,742  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              4                true                1024   bigint  avgt   50    22,488 ±  0,709  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              8               false                  32   bigint  avgt   50   364,359 ±  7,503  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              8               false                1024  varchar  avgt   50   377,809 ±  6,804  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              8               false                1024   bigint  avgt   50    46,098 ±  2,680  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              8                true                  32  varchar  avgt   50   616,763 ± 12,573  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              8                true                  32   bigint  avgt   50   394,600 ±  9,135  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              8                true                1024  varchar  avgt   50   157,529 ±  3,388  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              8                true                1024   bigint  avgt   50    37,747 ±  0,927  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             16               false                  32   bigint  avgt   50   676,770 ± 11,806  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             16               false                1024  varchar  avgt   50   698,540 ± 13,054  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             16               false                1024   bigint  avgt   50    80,182 ±  2,043  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             16                true                  32  varchar  avgt   50  1196,062 ± 26,167  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             16                true                  32   bigint  avgt   50   745,287 ± 13,933  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             16                true                1024  varchar  avgt   50   300,290 ±  5,760  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             16                true                1024   bigint  avgt   50    76,297 ±  1,689  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             32               false                  32   bigint  avgt   50  1380,569 ± 27,612  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             32               false                1024  varchar  avgt   50  1383,914 ± 34,170  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             32               false                1024   bigint  avgt   50   186,683 ±  4,778  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             32                true                  32   bigint  avgt   50  1525,892 ± 29,324  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             32                true                1024  varchar  avgt   50   607,039 ± 12,179  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             32                true                1024   bigint  avgt   50   153,029 ±  3,466  ms/op

AFTER:
Benchmark                                                      (columnCount)  (dictionaryBlocks)  (positionsPerPage)   (type)  Mode  Cnt     Score    Error  Units
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              2               false                  32  varchar  avgt   50   215,455 ±  4,112  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              2               false                  32   bigint  avgt   50    55,670 ±  1,539  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              2               false                1024  varchar  avgt   50   143,184 ±  2,665  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              2               false                1024   bigint  avgt   50    14,694 ±  0,368  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              2                true                  32  varchar  avgt   50   127,949 ±  2,723  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              2                true                  32   bigint  avgt   50    64,180 ±  1,563  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              2                true                1024  varchar  avgt   50    50,173 ±  1,268  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              2                true                1024   bigint  avgt   50    12,557 ±  0,320  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              4               false                  32  varchar  avgt   50   352,822 ±  8,106  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              4               false                  32   bigint  avgt   50    97,875 ±  2,084  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              4               false                1024  varchar  avgt   50   217,745 ±  4,598  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              4               false                1024   bigint  avgt   50    22,218 ±  0,591  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              4                true                  32  varchar  avgt   50   221,569 ±  4,802  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              4                true                  32   bigint  avgt   50   113,352 ±  2,705  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              4                true                1024  varchar  avgt   50    84,634 ±  1,600  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              4                true                1024   bigint  avgt   50    20,724 ±  0,636  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              8               false                  32   bigint  avgt   50   175,926 ±  3,526  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              8               false                1024  varchar  avgt   50   373,647 ±  8,506  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              8               false                1024   bigint  avgt   50    38,209 ±  0,945  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              8                true                  32  varchar  avgt   50   414,281 ± 11,566  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              8                true                  32   bigint  avgt   50   211,995 ±  4,899  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              8                true                1024  varchar  avgt   50   153,471 ±  2,961  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              8                true                1024   bigint  avgt   50    33,252 ±  1,145  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             16               false                  32   bigint  avgt   50   340,169 ±  6,795  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             16               false                1024  varchar  avgt   50   695,387 ± 13,538  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             16               false                1024   bigint  avgt   50    79,599 ±  5,121  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             16                true                  32  varchar  avgt   50   861,417 ± 15,827  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             16                true                  32   bigint  avgt   50   405,211 ±  8,541  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             16                true                1024  varchar  avgt   50   294,579 ±  5,309  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             16                true                1024   bigint  avgt   50    65,459 ±  2,010  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             32               false                  32   bigint  avgt   50   724,148 ± 13,855  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             32               false                1024  varchar  avgt   50  1389,267 ± 29,424  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             32               false                1024   bigint  avgt   50   170,175 ±  4,572  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             32                true                  32   bigint  avgt   50   881,287 ± 17,434  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             32                true                1024  varchar  avgt   50   582,909 ± 11,491  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             32                true                1024   bigint  avgt   50   133,659 ±  3,723  ms/op
```

All benchmarks improved. Some benchmarks improved significantly because of https://github.com/prestosql/presto/pull/467/commits/fda1a2a6523337c9efc7db3fd75bc6e3c7a2c1b4

dain commented on presto-main/src/main/java/io/prestosql/operator/WorkProcessorUtils.java:
```diff
@@ -224,20 +223,36 @@ else if (processor.isBlocked()) {
+            @Override
+            protected Void process(Optional<T> elementOptional)
+            {
+                if (elementOptional.isPresent()) {
```
I find the old lambda transform on the `elementOptional` more readable.

dain commented on presto-main/src/main/java/io/prestosql/operator/WorkProcessor.java:
```diff
@@ -267,90 +268,106 @@ R getResult()
          * Signals that process is blocked. {@link #process()} will be called again after {@code blocked} future is done.
          */
-        public static <T> ProcessState<T> blocked(ListenableFuture<?> blocked)
+        protected Void blocked(ListenableFuture<?> blocked)
```
Same question about `Void` vs `void`

sopel39 commented on presto-main/src/main/java/io/prestosql/operator/WorkProcessor.java:
```diff
@@ -260,41 +228,59 @@ private TransformationState(Type type, boolean needsMoreData, Optional<T> result
         {
-            return new TransformationState<>(Type.RESULT, needsMoreData, Optional.of(result), Optional.empty());
+            this.state = State.RESULT;
+            this.result = requireNonNull(result, "result is null");
```
I've ditched the approach with abstract `Transformation/Process` classes. Instead I have removed `Optional` boxing in `WorkProcessors` and it improved performance even better.

sopel39 commented:
@dain I've changed the approach for the last two commits and instead of using abstract classes I've removed `Optional` boxing in `WorkProcessor` which helped to improve performance a lot, here are the numbers:
```
After (no Optional)
Benchmark                                                      (columnCount)  (dictionaryBlocks)  (positionsPerPage)   (type)  Mode  Cnt     Score    Error  Units
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              2               false                  32  varchar  avgt   50   195,548 ±  1,666  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              2               false                  32   bigint  avgt   50    51,949 ±  0,975  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              2               false                1024  varchar  avgt   50   126,820 ±  0,937  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              2               false                1024   bigint  avgt   50    12,625 ±  0,296  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              2                true                  32  varchar  avgt   50   112,908 ±  2,078  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              2                true                  32   bigint  avgt   50    57,287 ±  0,532  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              2                true                1024  varchar  avgt   50    44,004 ±  0,267  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              2                true                1024   bigint  avgt   50    11,043 ±  0,054  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              4               false                  32  varchar  avgt   50   305,430 ±  1,557  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              4               false                  32   bigint  avgt   50    86,824 ±  0,724  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              4               false                1024  varchar  avgt   50   191,260 ±  1,034  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              4               false                1024   bigint  avgt   50    20,178 ±  0,961  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              4                true                  32  varchar  avgt   50   195,822 ±  2,005  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              4                true                  32   bigint  avgt   50    99,502 ±  0,998  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              4                true                1024  varchar  avgt   50    72,769 ±  0,341  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              4                true                1024   bigint  avgt   50    17,633 ±  0,254  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              8               false                  32   bigint  avgt   50   159,575 ±  2,967  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              8               false                1024  varchar  avgt   50   324,948 ±  1,997  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              8               false                1024   bigint  avgt   50    34,456 ±  1,614  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              8                true                  32  varchar  avgt   50   365,587 ±  3,016  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              8                true                  32   bigint  avgt   50   180,327 ±  1,383  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              8                true                1024  varchar  avgt   50   132,041 ±  0,760  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              8                true                1024   bigint  avgt   50    29,458 ±  0,562  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             16               false                  32   bigint  avgt   50   295,234 ±  3,455  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             16               false                1024  varchar  avgt   50   598,771 ±  4,379  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             16               false                1024   bigint  avgt   50    70,045 ±  4,179  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             16                true                  32  varchar  avgt   50   731,360 ±  4,786  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             16                true                  32   bigint  avgt   50   356,227 ±  3,698  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             16                true                1024  varchar  avgt   50   252,132 ±  1,542  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             16                true                1024   bigint  avgt   50    57,283 ±  1,162  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             32               false                  32   bigint  avgt   50   637,304 ±  5,081  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             32               false                1024  varchar  avgt   50  1190,130 ± 21,774  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             32               false                1024   bigint  avgt   50   146,601 ±  1,603  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             32                true                  32   bigint  avgt   50   753,605 ±  3,987  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             32                true                1024  varchar  avgt   50   504,934 ±  4,542  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             32                true                1024   bigint  avgt   50   115,600 ±  1,543  ms/op

After (abstract classes)
Benchmark                                                      (columnCount)  (dictionaryBlocks)  (positionsPerPage)   (type)  Mode  Cnt     Score    Error  Units
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              2               false                  32  varchar  avgt   50   215,455 ±  4,112  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              2               false                  32   bigint  avgt   50    55,670 ±  1,539  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              2               false                1024  varchar  avgt   50   143,184 ±  2,665  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              2               false                1024   bigint  avgt   50    14,694 ±  0,368  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              2                true                  32  varchar  avgt   50   127,949 ±  2,723  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              2                true                  32   bigint  avgt   50    64,180 ±  1,563  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              2                true                1024  varchar  avgt   50    50,173 ±  1,268  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              2                true                1024   bigint  avgt   50    12,557 ±  0,320  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              4               false                  32  varchar  avgt   50   352,822 ±  8,106  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              4               false                  32   bigint  avgt   50    97,875 ±  2,084  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              4               false                1024  varchar  avgt   50   217,745 ±  4,598  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              4               false                1024   bigint  avgt   50    22,218 ±  0,591  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              4                true                  32  varchar  avgt   50   221,569 ±  4,802  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              4                true                  32   bigint  avgt   50   113,352 ±  2,705  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              4                true                1024  varchar  avgt   50    84,634 ±  1,600  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              4                true                1024   bigint  avgt   50    20,724 ±  0,636  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              8               false                  32   bigint  avgt   50   175,926 ±  3,526  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              8               false                1024  varchar  avgt   50   373,647 ±  8,506  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              8               false                1024   bigint  avgt   50    38,209 ±  0,945  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              8                true                  32  varchar  avgt   50   414,281 ± 11,566  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              8                true                  32   bigint  avgt   50   211,995 ±  4,899  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              8                true                1024  varchar  avgt   50   153,471 ±  2,961  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented              8                true                1024   bigint  avgt   50    33,252 ±  1,145  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             16               false                  32   bigint  avgt   50   340,169 ±  6,795  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             16               false                1024  varchar  avgt   50   695,387 ± 13,538  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             16               false                1024   bigint  avgt   50    79,599 ±  5,121  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             16                true                  32  varchar  avgt   50   861,417 ± 15,827  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             16                true                  32   bigint  avgt   50   405,211 ±  8,541  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             16                true                1024  varchar  avgt   50   294,579 ±  5,309  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             16                true                1024   bigint  avgt   50    65,459 ±  2,010  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             32               false                  32   bigint  avgt   50   724,148 ± 13,855  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             32               false                1024  varchar  avgt   50  1389,267 ± 29,424  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             32               false                1024   bigint  avgt   50   170,175 ±  4,572  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             32                true                  32   bigint  avgt   50   881,287 ± 17,434  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             32                true                1024  varchar  avgt   50   582,909 ± 11,491  ms/op
BenchmarkScanFilterAndProjectOperator.benchmarkColumnOriented             32                true                1024   bigint  avgt   50   133,659 ±  3,723  ms/op
```
As one can see the performance with no `Optionals` is better in tight loops than with abstract classes.

kokosing commented:
@sopel39 Congrats! Nice.
